### PR TITLE
some more fixes for OneBoundedInterpreter replacement

### PR DESCRIPTION
**Pull request by rkuhn**

### CODE REVIEW

File: akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala [hunk 1]
```diff
@@ -41,6 +41,8 @@ class RecipeKeepAlive extends RecipeSpec {
 
       val subscription = sub.expectSubscription()
 
+      // FIXME RK: remove (because I think this cannot deterministically be tested and it might also not do what it should anymore)
```
Comment (rkuhn) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala]: @drewhk Please double check, but I don’t think this recipe makes sense any longer.

Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala]: This should be replaced by a keep-alive stage instead.

File: akka-http-core/src/main/scala/akka/http/impl/util/ByteReader.scala [hunk 2]
```diff
@@ -31,7 +31,7 @@ private[akka] class ByteReader(input: ByteString) {
     } else throw NeedMoreData
   def readShortLE(): Int = readByte() | (readByte() << 8)
   def readIntLE(): Int = readShortLE() | (readShortLE() << 16)
-  def readLongLE(): Long = (readIntBE() & 0xffffffffL) | ((readIntLE() & 0xffffffffL) << 32)
+  def readLongLE(): Long = (readIntLE() & 0xffffffffL) | ((readIntLE() & 0xffffffffL) << 32)
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/util/ByteReader.scala]: good un

File: akka-stream-testkit/src/main/scala/akka/stream/testkit/StreamTestKit.scala [hunk 3]
```diff
@@ -183,7 +184,14 @@ object TestSubscriber {
   final case class OnSubscribe(subscription: Subscription) extends SubscriberEvent
   final case class OnNext[I](element: I) extends SubscriberEvent
   final case object OnComplete extends SubscriberEvent
-  final case class OnError(cause: Throwable) extends SubscriberEvent
+  final case class OnError(cause: Throwable) extends SubscriberEvent {
```
Comment (drewhk) [inline akka-stream-testkit/src/main/scala/akka/stream/testkit/StreamTestKit.scala]: another nice one!

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala [hunk 4]
```diff
@@ -104,11 +104,9 @@ class FlowMapAsyncSpec extends AkkaSpec {
         n
       }).to(Sink(c)).run()
       val sub = c.expectSubscription()
-      // running 8 in parallel
-      probe.receiveN(8).toSet should be((1 to 8).toSet)
       probe.expectNoMsg(500.millis)
       sub.request(1)
-      probe.expectMsg(9)
+      probe.receiveN(9).toSet should be((1 to 9).toSet)
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala]: why was this change necessary?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala [hunk 5]
```diff
@@ -73,7 +73,7 @@ class SourceSpec extends AkkaSpec {
   }
 
   "Maybe Source" must {
-    "complete materialized future with None when stream cancels" in {
+    "complete materialized future with None when stream cancels" in Utils.assertAllStagesStopped {
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceSpec.scala]: nice, thanks

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 6]
```diff
@@ -537,149 +537,146 @@ private[akka] object MapAsync {
...
-    else ctx.finish()
+    setHandler(out, new OutHandler {
+      override def onPull(): Unit = {
+        if (!hasBeenPulled(in)) tryPull(in)
+        if (!buffer.isEmpty) push(out, buffer.dequeue())
+        else if (isClosed(in) && todo == 0) completeStage()
+      }
+    })
+  }
 }
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: can we remove AsyncStage now or are there any other implementations?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 7]
```diff
@@ -173,15 +173,12 @@ class MergePreferred[T] private (val secondaryPorts: Int, val eagerClose: Boolea
       else tryPull(in)
     }
 
-    // FIXME: slow iteration, try to make in a vector and inject into shape instead
-    (0 until secondaryPorts).map(in).foreach { i ⇒
+    shape.inSeq.foreach { i ⇒
       setHandler(i, new InHandler {
         override def onPush(): Unit = {
           if (isAvailable(out)) {
-            if (!pending) {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: yep, this was not removed before.

**Comment by drewhk** [comment]
LGTM

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala]: because both mapAsync and mapAsyncUnordered now require demand in order to start running

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 6 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: have not checked yet, will do!

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: good point: there’s just one more special-case test utility that needs replacement :-)

**Comment by rkuhn** [comment]
… now with websocket support; we should probably merge after review and before I add more things to it ;-)

**Comment by rkuhn** [comment]
@drewhk merge at will, I’ll open a new PR afterwards including possible review comment fixes

**Comment by drewhk** [comment]
Ok, I will :)
